### PR TITLE
Added an affirmative twig grant checker

### LINKED ISSUE #5129: Dashboard BLOCKS roles are treated with "AND"
#### Sonata packages

```
sonata-project/admin-bundle              3.x-dev 82a4d6e dev-master 4b2e943 The missing Symfony Admin Generator
sonata-project/block-bundle              3.12.1          3.12.1             Symfony SonataBlockBundle
sonata-project/cache                     2.0.1           2.0.1              Cache library
sonata-project/core-bundle               3.9.1           3.10.0             Symfony SonataCoreBundle
sonata-project/datagrid-bundle           2.3.1           2.3.1              Symfony SonataDatagridBundle
sonata-project/doctrine-orm-admin-bundle 3.x-dev ec1d2a6 dev-master 1468a90 Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/exporter                  1.9.0           1.9.0              Lightweight Exporter library
```

#### Symfony packages

```
Too much of it, 4.0.11 basically.
```

#### PHP version

```
PHP 7.2.6
```

## Subject

Dashboard groups with roles are treated with "AND"

## Steps to reproduce

```yaml
sonata_admin:
  dashboard:
    blocks:
      - position: top
        type: sonata.admin.block.stats
        class: col-lg-3 col-xs-6
        settings:
          code: app.admin.employee
          icon: fa-users
          text: app.admin.employees_enabled
          color: bg-blue
          filters:
            enabled: { value: 1 }
        roles: [ROLE_SYS_ADMIN, ROLE_CLIENT_ADMIN]
```

## Expected results

Both ROLE_SYS_ADMIN and ROLE_CLIENT_ADMIN see the stats block.

## Actual results

Only a user with both roles sees the stats block.

*NOT* caused by https://github.com/sonata-project/SonataAdminBundle/pull/5086 and thus not related to #5128

### PR DESCRIPTION
**Pull request by andrewmy**
I am targeting this branch, because it's a fix.

Closes #5129

## Changelog

```markdown
### Added
- An affirmative grant checker for Twig

### Fixed
- Dashboard block security was expected to be checked affirmatively rather than unanimously
```
## Subject

Dashboard block security was expected to be checked affirmatively rather than unanimously, fixed it by adding a new checker function to Twig.

### CODE REVIEW

**Comment by andrewmy** [comment]
Now this actually might be a BC change, as opposed to #5130.

**Comment by kunicmarko20** [comment]
I think it is not a break.

**Comment by kunicmarko20** [comment]
Is it a patch or minor @OskarStark, I mean he added new feature, or I am wrong? cc @greg0ire

**Comment by andrewmy** [comment]
Re: Travis builds, why doesn't it see `security.authorization_checker`?

File: src/Twig/Extension/SonataAdminExtension.php [hunk 1]
```diff
@@ -501,6 +512,41 @@ public function getXEditableChoices(FieldDescriptionInterface $fieldDescription)
...
+            $object = new FieldVote($object, $field);
+        }
+
+        if (!is_array($role)) {
+            $role = [$role];
+        }
+
+        foreach ($role as $oneRole) {
+            try {
+                $grant = $this->securityChecker->isGranted($oneRole, $object);
```
Comment (greg0ire) [inline src/Twig/Extension/SonataAdminExtension.php]:
```php
<?php
if ($this->securityChecker->isGranted($oneRole, $object)) {
    return true;
}
```

Comment (OskarStark) [reply src/Twig/Extension/SonataAdminExtension.php]: No, but a `return false;` in the catch block, right @greg0ire ?

Comment (OskarStark) [reply src/Twig/Extension/SonataAdminExtension.php]: Ah no wait, return false is not possible at this point 😄 🤔

**Comment by andrewmy** [comment]
Ready to merge?

**Comment by andrewmy** [comment]
Travis fails with this message:

> The service "sonata.admin.twig.extension" has a dependency on a non-existent service "security.authorization_checker".

At the same time the `security.authorization_checker` service is used as a dependency for services in `menu.xml` and `security.xml`. Any idea why the auth checker is not recognized here of all places?

**Comment by OskarStark** [comment]
Any idea @greg0ire ?

**Comment by kunicmarko20** [comment]
And green, 😄

**Comment by kunicmarko20** [comment]
Np, glad we get to merge this!

**Comment by greg0ire** [comment]
Please fix the cs issue though

**Comment by andrewmy** [comment]
#pedantic 😂
I didn't touch the file at all and FlintCI was happy with it. Ok, fixed it.

**Comment by kunicmarko20** [comment]
that was fixed yesterday by you @greg0ire, maybe just a rebase is needed?

**Comment by greg0ire** [comment]
Please kindly squash your commits together. If you don't, we'll try to remember to do it for you but it's best if you save us this trouble.

How to do that?

1. `git rebase -i origin/3.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. Close your editor, git should do its magic, and you should end up with one commit
4. Use `git push --force` to overwrite what you already pushed. Don't forget the `--force` option otherwise git will try to merge both things together.

**Comment by greg0ire** [comment]
> Dashboard block security was expected to be checked affirmatively rather than unanimously,

Is it? Why? Please state the obvious.

**Comment by andrewmy** [comment]
tl;dr: When defining a dashboard block with `roles: [ROLE_ADMIN_ONE, ROLE_ADMIN_TWO]`, you would expect it to show if the user has either of those roles.

**Comment by greg0ire** [comment]
Now I get it!

**Comment by andrewmy** [comment]
Merge? :D

**Comment by kunicmarko20** [comment]
Thank you @andrewmy !
